samplejava/src/main/java/com/example/chattutorialjava/ChannelActivity3.java

```diff
@@ -45,100 +45,91 @@ public static Intent newIntent(Context context, Channel channel) {
-                        User user = ((TypingStopEvent) event).getUser();
-                        String name = (String) user.getExtraData().get("name");
-                        Set<String> typingCopy = currentlyTyping.getValue();
-                        typingCopy.remove(name);
-                        currentlyTyping.postValue(typingCopy);
-                    }
-                    return null;
-                });
-        currentlyTyping.observe(this, users -> {
-            String typing = "nobody is typing";
-            if (!users.isEmpty()) {
-                typing = "typing: " + TextUtils.join(", ", users);
+        // Step 4 - Handle navigate up state
+        messageListViewModel.getState().observe(this, state -> {
+            if (state instanceof NavigateUp) {
+                finish();
             }
-            channelHeaderView.setText(typing);
->>>>>>> 23067e524310df9e383d921476e1f7625c2b04df
         });
 
-        final MessageInputViewModel messageInputViewModel = viewModelProvider.get(MessageInputViewModel.class);
-        MessageInputViewModelBinding.bind(messageInputViewModel, messageInputView, this);
-        messageListViewModel.getMode().observe(this, mode -> {
-            if (mode instanceof MessageListViewModel.Mode.Thread) {
-                messageInputViewModel.setActiveThread(((MessageListViewModel.Mode.Thread) mode).getParentMessage());
-            } else if (mode instanceof MessageListViewModel.Mode.Normal) {
-                messageInputViewModel.resetThread();
-            }
-            messageListView.setOnMessageEditHandler((message) -> {
-                messageInputViewModel.getEditMessage().postValue(message);
-                return Unit.INSTANCE;
-            });
+        // Step 5 - Let the message input know when we are editing a message
+        messageListView.setOnMessageEditHandler(message -> {
+            messageInputViewModel.getEditMessage().postValue(message);
+            return Unit.INSTANCE;
         });
 
-        Function0<Unit> onBackClick = () -> {
+        // Step 6 - Handle back button behaviour correctly when you're in a thread
+        channelHeaderView.setOnBackClick(() -> {
             messageListViewModel.onEvent(MessageListViewModel.Event.BackButtonPressed.INSTANCE);
             return Unit.INSTANCE;
-        };
-
+        });
         getOnBackPressedDispatcher().addCallback(this, new OnBackPressedCallback(true) {
             @Override
             public void handleOnBackPressed() {
-                onBackClick.invoke();
+                channelHeaderView.getOnBackClick().invoke();
+            }
+        });
+
+        // Custom typing info header bar
+        String nobodyTyping = "nobody is typing";
+        typingHeader.setText(nobodyTyping);
+
+        // Obtain a ChannelController
+        ChannelController channelController =
+                ChatDomain.instance().getUseCases()
+                        .getWatchChannel().invoke(cid, 0)
```

Comment:
We should replace `getWatchChannel` with `getChannelController` after next release. Maybe let's add TODO?